include/secp256k1_musig.h

```diff
@@ -197,155 +154,120 @@ SECP256K1_API SECP256K1_WARN_UNUSED_RESULT int secp256k1_musig_pubkey_tweak_add(
+/** Takes the public nonces of all signers and computes a session cache that is
+ *  required for signing and verification of partial signatures and a signature
+ *  template that is required for combining partial signatures.
  *
- *  Returns: 1: nonces are successfully combined
- *           0: a signer's nonce is missing
- *  Args:        ctx: pointer to a context object (cannot be NULL)
- *           session: session to update with the combined public nonce (cannot be
- *                    NULL)
- *           signers: an array of signers' data, which must have had public nonces
- *                    set with `musig_set_nonce`. Array length must equal to `n_signers`
- *                    (cannot be NULL)
- *         n_signers: the length of the signers array. Must be the total number of
- *                    signers participating in the MuSig.
- *  Out: nonce_parity: if non-NULL, a pointer to an integer that indicates the
- *                    parity of the combined public nonce. Used for adaptor
- *                    signatures.
- *           adaptor: point to add to the combined public nonce. If NULL, nothing is
- *                    added to the combined nonce.
+ *  Returns: 0 if the arguments are invalid or if all signers sent invalid
+ *           pubnonces, 1 otherwise
+ *  Args:         ctx: pointer to a context object, initialized for verification
+ *                     (cannot be NULL)
+ * Out: session_cache: pointer to a struct to store the session_cache, which is
+ *                     used for partial_sign and partial_verify
+ *       sig_template: pointer to a struct to store a sig template, which is
+ *                     used for partial_sig_combine
+ *       nonce_parity: optional pointer to an integer that indicates the parity
+ *                     of the combined public nonce. Used for adaptor
+ *                     signatures. (can be NULL)
+ * In:      pubnonces: array of pointers to the 66-byte pubnonces sent by the
+ *                     signers
+ *        n_pubnonces: number of elements in the pubnonces array. Must be
+ *                     greater than 0.
+ *              msg32: the 32-byte message to sign
+ *        combined_pk: pointer to combined public key of all signers
+ *        pre_session: pointer to the pre_session that was output when
+ *                     combined_pk was created
+ *            adaptor: optional pointer to an adaptor if this signing session is
+ *                     part of an adaptor signature protocol (can be NULL)
  */
-SECP256K1_API int secp256k1_musig_session_combine_nonces(
+SECP256K1_API int secp256k1_musig_process_nonces(
     const secp256k1_context* ctx,
-    secp256k1_musig_session *session,
-    const secp256k1_musig_session_signer_data *signers,
-    size_t n_signers,
+    secp256k1_musig_session_cache *session_cache,
+    secp256k1_musig_template *sig_template,
     int *nonce_parity,
+    const unsigned char * const* pubnonces,
```

Comment:
To make the API easier to understand, I would advice to remove `pubnonces` and `n_pubnonces`, and only accept `combined_pubnonce`.

Reply:
That's fair, but then everyone has to call `musig_nonces_combine` instead of allowing signers without a third-party nonce aggregator to skip this function altogether.

Reply:
If you decide to keep `pubnonces` here, consider using a [strong type](https://github.com/ElementsProject/secp256k1-zkp/pull/131/files#r691687312)